commercelayer/rate_limiter.go

```diff
@@ -0,0 +1,43 @@
+package commercelayer
+
+import (
+	"net/http"
+	"sync"
+	"time"
+)
+
+// throttledTransport embeds the underlying transport
+// and regulates requests to respect rate limits.
+type throttledTransport struct {
+	transport http.RoundTripper
+	mutex     sync.Mutex
+}
+
+// RoundTrip executes an HTTP request, and if the response indicates rate
+// limiting (HTTP 429), it waits for the retry interval specified in the
+// "X-Ratelimit-Interval" header before retrying. It locks a mutex to ensure
+// only one request is processed at a time.
+func (t *throttledTransport) RoundTrip(req *http.Request) (*http.Response, error) {
```

Comment:
This looks good! I would just make the rate limiting logic work by default, so no need for the disable/enable option

Reply:
I am not sure it is a good idea to enable it by default, as the mutex creates a constriction effect.

Terraform has got parallelism as a first-class principle. When possible, it will use the provider and send requests multiple requests in parallel. The implementation I provide here will enforce one request at a time with this general mutex.

https://developer.hashicorp.com/terraform/cli/commands/apply#parallelism-n

When using this provider to heavily modify data (with big payloads), the rate-limiting will often be a problem so in that context, having it by default would make sense.

But for smaller payloads, having it disabled will execute way faster as requests will be parallelized without hitting rate limits.

What do you think?

Reply:
I think the main goal should be safety here - so I think I would prefer to always make the rate limiting logic work, just like CommerceLayer does it in their documented `retrieveAll` function from the SDK utils.

Reply:
I just pushed a new commit to enable this rate-limiting mechanism by default, with still the possibility to disable it if the user wants that.